pkg/cli/admin/mustgather/mustgather.go

```diff
@@ -412,11 +502,16 @@ func (o *MustGatherOptions) Run() error {
 	defer o.logTimestamp()
 
 	var wg sync.WaitGroup
-	wg.Add(len(pods))
+	w := make(chan struct{}, concurrentMG)
```

Comment:
Why do we need this?. Are we running 4 must-gather pods or are we running 4 different processes in the same must-gather pod?. I think, we can achieve by only using `wg`

Reply:
This is because we want to limit the must-gather to startup only 4 pods at a time. https://github.com/openshift/enhancements/pull/1487#discussion_r1402228181

Reply:
@ardaguclu can you please mark this bit resolved?

Reply:
wait group is incremented by the number of pods(which is 4) and we don't need additional channel(`w`)

Reply:
We just need a `wg.Wait`

Reply:
> wait group is incremented by the number of pods(which is 4) and we don't need additional channel(`w`)

Don't we need the channel to ensure that there are at most 4 pods running at any given time? 🤔 

How do you suggest we ensure that otherwise?

Reply:
@ardaguclu tagging you just in case this slipped

Reply:
@ardaguclu PTAL https://go.dev/play/p/r8IjXNTNtfR. Please run it with and without the channel and let me know your thoughts.

Reply:
hmm, now I got the point. Thanks. I wonder is there any simpler solution to always have 4 pods in any time frame because this mechanism looks to me hard to read.

Reply:
[semaphore package](https://pkg.go.dev/golang.org/x/sync/semaphore), maybe. But I, personally, find this simpler than using the semaphore package.

Reply:
Do you think adding a code comment would help? I recently did so for the two new constants being introduced in this PR.

Reply:
If it is ok for you I think we can try with semaphore package but I'm not sure as well.

Reply:
TBH, I'm OK to try it since it'll be a new thing to learn for me. 😉 

However, I've observed more use of this pattern vs. semaphore package. Semaphore package, I think, shines when we need weighted bounded concurrency. Ours is a use case of bounded concurrency.

If you don't mind, I'd like @soltysh & @ingvagabund to chime in before we decide.

Reply:
Alternatively, you can create a queue of pods and have concurrentMG number of workers to process the items in the queue. Which is a well-establish solution and produces a nice looking code. E.g. https://github.com/openshift/library-go/blob/master/pkg/controller/factory/base_controller.go#L109-L119 and https://github.com/openshift/library-go/blob/master/pkg/controller/factory/base_controller.go#L246.

Reply:
@ingvagabund not gonna lie, I understand the worker pool pattern, but I'm a bit hung here.
Do you want me to use the `baseController` implementation that you've linked to, or can I try doing something simpler?
The `baseController` implementation isn't making complete sense to me. Especially, how is it limiting the workers to a certain number. Hence, I'm hung.

Reply:
It was an example. Any worker pool pattern works.

Reply:
> I was unable to find how/if/where this limit was being set.

passed through `workers` variable

> Looking at baseController it looks like a way to implement/write k8s controllers. I don't think it needs a way to limit workers built into it.

Only the queue and worker loop from the base controller is relevant.

Reply:
Marking this long discussion resolved since worker pool pattern is implemented. Request you to start a new discussion if you have any feedback around this part of the code. 😆